### PR TITLE
Aws sts

### PR DESCRIPTION
**Pull request by EItanya**
# Description

Support assuming AWS iam roles for lambda requests via the AWS STS AssumeRoleWithWebToken API.

# Context

This feature allows Gloo instances running in EKS to use AWS Service Account credentials to authenticate requests directly.

# Checklist:

- [x] I included a concise, user-facing changelog (for details, see https://github.com/solo-io/go-utils/tree/master/changelogutils) which references the issue that is resolved.
- [x] If I updated APIs (our protos) or helm values, I ran `make install-go-tools generated-code` to ensure there will be no code diff
- [x] I followed guidelines laid out in the Gloo [contribution guide](https://docs.solo.io/gloo/latest/contributing/)
- [x] I opened a draft PR or added the work in progress label if my PR is not ready for review
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works
BOT NOTES: 
resolves https://github.com/solo-io/gloo/issues/3309

### CODE REVIEW

File: projects/gloo/pkg/plugins/aws/plugin_test.go [hunk 1]
```diff
@@ -296,9 +298,85 @@ var _ = Describe("Plugin", func() {
...
+	})
+
+	Context("service account creds", func() {
+
+		var (
+			cfg *AWSLambdaConfig
+
+			saCredentials = &AWSLambdaConfig_ServiceAccountCredentials{
+				Cluster: "aws_sts",
+				Uri:     "sts.aws.com",
```
Comment (yuval-k) [inline projects/gloo/pkg/plugins/aws/plugin_test.go:331]: `sts.amazonaws.com`?

File: projects/gloo/pkg/plugins/aws/plugin_test.go [hunk 2]
```diff
@@ -296,9 +298,85 @@ var _ = Describe("Plugin", func() {
...
+				},
+			}
+
+			roleArn = "role_arn"
+		)
+
+		BeforeEach(func() {
+			cfg = &AWSLambdaConfig{}
+
+			awsPlugin.Init(plugins.InitParams{
```
Comment (yuval-k) [inline projects/gloo/pkg/plugins/aws/plugin_test.go:344]: can we use settings?
see here:
https://github.com/solo-io/gloo/blob/eda9aa787cb54399fdef5b24c597088d77b279e5/test/e2e/custom_auth_test.go#L79

File: projects/gloo/api/external/envoy/extensions/aws/filter.proto [hunk 3]
```diff
@@ -43,15 +49,48 @@ message AWSLambdaProtocolExtension {
...
+    // If credentials are provided on the cluster (using the
+    // AWSLambdaProtocolExtension), it will override these credentials. This
+    // defaults to false, but may change in the future to true.
+    google.protobuf.BoolValue use_default_credentials = 1;
+
+    // Use projected service account token, and role arn to create reate temporary
+    // credentials with which to authenticate lambda requests.
+    // This functionality is meant to work along side EKS service account to IAM
+    // binding functionality as outlined here:
+    // https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.htmll
```
Comment (kdorosh) [inline projects/gloo/api/external/envoy/extensions/aws/filter.proto:72]: broken link,
```suggestion
    // https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html
```

File: projects/gloo/api/external/envoy/extensions/aws/filter.proto [hunk 4]
```diff
@@ -43,15 +49,48 @@ message AWSLambdaProtocolExtension {
...
+    //
+    // If the following environment values are not present, this option cannot be used.
+    //   1. AWS_WEB_IDENTITY_TOKEN_FILE
+    //   2. AWS_ROLE_ARN
```
Comment (kdorosh) [inline projects/gloo/api/external/envoy/extensions/aws/filter.proto:76]: what happens if i try to do this? as a user, if I were to misconfigure, it would be nice to know what to look for

File: projects/gloo/pkg/plugins/aws/plugin.go [hunk 5]
```diff
@@ -227,12 +230,28 @@ func (p *plugin) ProcessRoute(params plugins.RouteParams, in *v1.Route, out *env
...
+		}
+	case *v1.GlooOptions_AWSOptions_ServiceAccountCredentials:
+		filterconfig.CredentialsFetcher = &AWSLambdaConfig_ServiceAccountCredentials_{
+			ServiceAccountCredentials: &AWSLambdaConfig_ServiceAccountCredentials{
+				Cluster: typedFetcher.ServiceAccountCredentials.GetCluster(),
+				Uri:     typedFetcher.ServiceAccountCredentials.GetUri(),
+				Timeout: typedFetcher.ServiceAccountCredentials.GetTimeout(),
+			},
+		}
+	}
```
Comment (kdorosh) [inline projects/gloo/pkg/plugins/aws/plugin.go:254]: why not do:
```golang
	case *v1.GlooOptions_AWSOptions_ServiceAccountCredentials:
		filterconfig.CredentialsFetcher = &AWSLambdaConfig_ServiceAccountCredentials_{
			ServiceAccountCredentials:typedFetcher.ServiceAccountCredentials,
		}
	}
```

would be shorter and more future-proof

File: projects/gloo/pkg/plugins/aws/plugin_test.go [hunk 6]
```diff
@@ -296,9 +298,85 @@ var _ = Describe("Plugin", func() {
 
 			process()
 
-			Expect(cfg.UseDefaultCredentials.GetValue()).To(BeTrue())
+			Expect(cfg.GetUseDefaultCredentials().GetValue()).To(BeTrue())
+			Expect(lpe.AccessKey).To(Equal(accessKeyValue))
+			Expect(lpe.SecretKey).To(Equal(secretKeyValue))
```
Comment (kdorosh) [inline projects/gloo/pkg/plugins/aws/plugin_test.go:303]: should we be asserting that the session token is empty here?

File: test/e2e/aws_test.go [hunk 7]
```diff
@@ -359,4 +368,123 @@ var _ = Describe("AWS Lambda", func() {
...
+		addCredentialsSts := func() {
+
+			webToken := os.Getenv(webIdentityToken)
+			if webToken == "" {
+				Skip(fmt.Sprintf("Token location unset, set via %s", webIdentityToken))
+			}
+
+			roleArn := os.Getenv(awsRoleArnSts)
+			if roleArn == "" {
+				Skip(fmt.Sprintf("AWS role arn unset, set via %s", awsRoleArnSts))
```
Comment (kdorosh) [inline test/e2e/aws_test.go:386]: should this be a `Fail` instead? in case we accidentally remove the env vars from CI

File: projects/gloo/pkg/plugins/aws/plugin_test.go (hunk 2 above)
Comment (EItanya) [reply projects/gloo/pkg/plugins/aws/plugin_test.go:344]: This isn't an e2e test, don't have `RunGlooGateway`

File: test/e2e/aws_test.go (hunk 7 above)
Comment (EItanya) [reply test/e2e/aws_test.go:386]: The rest of the tests were set to `Skip` Happy to switch to `Fail`

File: projects/gloo/api/external/envoy/extensions/aws/filter.proto (hunk 4 above)
Comment (EItanya) [reply projects/gloo/api/external/envoy/extensions/aws/filter.proto:76]: Envoy will NACK the configuration, so in envoy-gloo-ee the stat will show up, in envoy-gloo it will show up in the logs

File: projects/gloo/pkg/plugins/aws/plugin_test.go (hunk 6 above)
Comment (EItanya) [reply projects/gloo/pkg/plugins/aws/plugin_test.go:304]: No, the secret data will still be populated to maintain backwards compatibility

File: install/helm/gloo/generate/values.go [hunk 8]
```diff
@@ -116,6 +116,13 @@ type Settings struct {
...
 	DisableKubernetesDestinations bool                 `json:"disableKubernetesDestinations" desc:"Gloo allows you to directly reference a Kubernetes service as a routing destination. To enable this feature, Gloo scans the cluster for Kubernetes services and creates a special type of in-memory Upstream to represent them. If the cluster contains a lot of services and you do not restrict the namespaces Gloo is watching, this can result in significant overhead. If you do not plan on using this feature, you can set this flag to true to turn it off."`
+	Aws                           AwsSettings          `json:"aws,omitempty"`
+}
+
+type AwsSettings struct {
+	EnableCredentialsDiscovery      bool   `json:"enableCredentialsDiscovery" desc:"Enable AWS credentials discovery in Envoy for lambda requests."`
```
Comment (yuval-k) [inline install/helm/gloo/generate/values.go:123]: EnableCredentialsDiscovery and EnableServiceAccountCredentials are mutually exclusive; IIRC we had a plan to change that; but for now let's mention that and which one takes precedence.

File: projects/gloo/api/external/envoy/extensions/aws/filter.proto [hunk 9]
```diff
@@ -43,15 +49,51 @@ message AWSLambdaProtocolExtension {
...
+
+    // Use projected service account token, and role arn to create reate temporary
+    // credentials with which to authenticate lambda requests.
+    // This functionality is meant to work along side EKS service account to IAM
+    // binding functionality as outlined here:
+    // https://docs.aws.amazon.com/eks/latest/userguide/iam-roles-for-service-accounts.html
+    //
+    // If the following environment values are not present, this option cannot be used.
+    //   1. AWS_WEB_IDENTITY_TOKEN_FILE
+    //   2. AWS_ROLE_ARN
```
Comment (yuval-k) [inline projects/gloo/api/external/envoy/extensions/aws/filter.proto:76]: are these docs user facing? if so we should mention that these should be set on the gateway-proxy deployment (and not gloo)

File: projects/gloo/api/v1/options/aws/aws.proto [hunk 10]
```diff
@@ -30,6 +30,10 @@ message UpstreamSpec {
     // The list of Lambda Functions contained within this region.
     // This list will be automatically populated by Gloo if discovery is enabled for AWS Lambda Functions
     repeated LambdaFunctionSpec lambda_functions = 3;
+
+    // (Optional): role_arn to use when assuming a role for a given request via STS.
+    // This option will only be respected if STS credentials are enabled.
```
Comment (yuval-k) [inline projects/gloo/api/v1/options/aws/aws.proto:35]: help the user understand what does "This option will only be respected if STS credentials are enabled"
perhaps point to the field in the settings we need to change?
mention that this overrides the default one in the settings if provided

File: projects/gloo/api/external/envoy/extensions/aws/filter.proto (hunk 4 above)
Comment (EItanya) [reply projects/gloo/api/external/envoy/extensions/aws/filter.proto:76]: Adding this to a comment

File: projects/gloo/api/v1/settings.proto [hunk 11]
```diff
@@ -335,12 +336,20 @@ message GlooOptions {
...
+            // on the upstream when running on AWS environment.
+            //
+            // Note: This should **ONLY** be enabled when running in an AWS environment, as the AWS
+            // code blocks the envoy main thread. This should be negligible when running inside AWS.
+            bool enable_credentials_discovey = 1;
+
+            // In order to specify the aws sts endpoint, both the cluster and uri must be set.
+            // This is due to an envoy limitation which cannot infer the host or path from the cluster,
+            // and therefore must be explicitly specified via the uri
+            envoy.config.filter.http.aws_lambda.v2.AWSLambdaConfig.ServiceAccountCredentials service_account_credentials = 2;
```
Comment (yuval-k) [inline projects/gloo/api/v1/settings.proto:351]: we should mention which takes precedence if both are set.
as these are the main user facing docs, i would also mentioned the requirements for the env var in the gateway proxy deployment (and mention that the role as optional can be provided in the upstream).

File: projects/gloo/api/v1/settings.proto [hunk 12]
```diff
@@ -335,12 +336,20 @@ message GlooOptions {
...
-        // code blocks the envoy main thread. This should be negligible when running inside AWS.
-        bool enable_credentials_discovey = 1;
+
+        oneof credentials_fetcher {
+            // Enable credential discovery via IAM; when this is set, there's no need provide a secret
+            // on the upstream when running on AWS environment.
+            //
+            // Note: This should **ONLY** be enabled when running in an AWS environment, as the AWS
+            // code blocks the envoy main thread. This should be negligible when running inside AWS.
+            bool enable_credentials_discovey = 1;
```
Comment (yuval-k) [inline projects/gloo/api/v1/settings.proto:346]: i think its too late but:
```
enable_credentials_discovey -> enable_credentials_discovery
```
(typo in discovery - missing r)

File: test/e2e/aws_test.go [hunk 13]
```diff
@@ -278,7 +287,7 @@ var _ = Describe("AWS Lambda", func() {
 			localAwsCredentials := credentials.NewSharedCredentials("", "")
 			v, err := localAwsCredentials.Get()
 			if err != nil {
-				Skip("no AWS creds available")
+				Fail("no AWS creds available")
```
Comment (yuval-k) [inline test/e2e/aws_test.go:290]: why this change?

File: projects/gloo/api/v1/settings.proto (hunk 12 above)
Comment (kdorosh) [reply projects/gloo/api/v1/settings.proto:346]: yeah, too late unless we want to fix the spelling and deprecate the old one

File: test/e2e/aws_test.go (hunk 13 above)
Comment (EItanya) [reply test/e2e/aws_test.go:290]: Kevin mentioned [above](https://github.com/solo-io/gloo/pull/3452#discussion_r464643781) that it would be good to fail if credentials aren't available in CI

File: projects/gloo/api/v1/settings.proto (hunk 12 above)
Comment (EItanya) [reply projects/gloo/api/v1/settings.proto:346]: yes, would be a break to fix it

File: projects/gloo/api/v1/settings.proto (hunk 11 above)
Comment (EItanya) [reply projects/gloo/api/v1/settings.proto:351]: How can both be set, it's a oneof.
